dali/operators/image/crop/bbox_crop.cc

```diff
@@ -14,143 +14,655 @@
 
+#include <algorithm>
+#include <random>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include "dali/core/common.h"
+#include "dali/core/error_handling.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/pipeline/util/batch_rng.h"
+#include "dali/pipeline/util/bounding_box.h"
+
 namespace dali {
 
+namespace {
+
+// This is the default bbox layout that the operator uses internally
+TensorLayout InternalBboxLayout(int ndim, bool ltrb) {
+  assert(ndim == 3 || ndim == 2);
+  if (ltrb) {
+    return ndim == 3 ? "xyzXYZ" : "xyXY";
+  } else {
+    return ndim == 3 ? "xyzWHD" : "xyWH";
+  }
+}
+
+// This is the default shape layout that the operator uses internally
+TensorLayout InternalShapeLayout(int ndim) {
+  assert(ndim == 3 || ndim == 2);
+  return ndim == 3 ? "WHD" : "WH";
+}
+
+void CollectShape(std::vector<TensorShape<>> &v,
+                  const std::string &name,
+                  const OpSpec& spec,
+                  const workspace_t<CPUBackend>& ws,
+                  int ndim,
+                  bool required = false) {
+  int batch_size = spec.GetArgument<int>("batch_size");
+  v.clear();
+  v.reserve(batch_size);
+
+  if (spec.HasTensorArgument(name)) {
+    auto arg_view = view<const int>(ws.ArgumentInput(name));
+    DALI_ENFORCE(arg_view.num_samples() == batch_size, make_string(
+      "Unexpected number of samples in argument `", name, "`: ", arg_view.num_samples(),
+      ", expected: ", batch_size));
+    DALI_ENFORCE(arg_view.sample_dim() == ndim, make_string(
+      "Unexpected number of dimensions in argument `", name, "`: ", arg_view.sample_dim(),
+      ", expected: ", ndim));
+
+    std::vector<int64_t> tmp(ndim);
+    for (int sample = 0; sample < batch_size; sample++) {
+      const auto* sample_data = arg_view.tensor_data(sample);
+      for (int d = 0; d < ndim; d++) {
+        tmp[d] = static_cast<int64_t>(sample_data[d]);
+      }
+      v.emplace_back(tmp);
+    }
+  } else if (spec.HasArgument(name)) {
+    auto tmp = spec.GetRepeatedArgument<int>(name);
+    DALI_ENFORCE(static_cast<int>(tmp.size()) == ndim,
+      make_string("Argument `", name, "` must be a ", ndim, "D vector"));
+
+    TensorShape<> sh(std::vector<int64_t>(tmp.begin(), tmp.end()));
+    v.resize(batch_size, sh);
+  } else {
+    if (required)
+      DALI_FAIL(make_string("Argument `", name, "` is required"));
+  }
+}
+
+struct SampleOption {
+  bool no_crop = false;
+  float min_iou = 0.0f;
+  SampleOption(bool _no_crop, float _min_iou)
+      : no_crop(_no_crop), min_iou(_min_iou) {}
+};
+
+struct Range {
+  Range(float _min, float _max)
+      : min(_min), max(_max) {
+    DALI_ENFORCE(min >= 0.0f,
+      make_string("Min should be at least 0.0. Received: ", min));
+    DALI_ENFORCE(min <= max,
+      make_string("Range should be provided as: [min, max]. Received: [", min, max, "]"));
+  }
+
+  explicit Range(const std::vector<float>& range)
+      : Range(range.size() == 2 ? range[0] : -1.0f,
```

Comment:
Is `range.size() == 2` allowed or it is an error?

Reply:
removed